src/validation.cpp

```diff
@@ -2379,7 +2429,7 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
 
         // Connect new blocks.
         BOOST_REVERSE_FOREACH(CBlockIndex *pindexConnect, vpindexToConnect) {
-            if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : std::shared_ptr<const CBlock>(), connectTrace)) {
+            if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : std::shared_ptr<const CBlock>(), connectTrace, disconnectpool)) {
```

Comment:
@TheBlueMatt I pushed up a commit that addresses the edge-cases better, I think -- it now will just erase from the mempool to become consistent again.  Also patches up the missing `mempool.removeForReorg` calls in a couple places, and moves all the reorg consistency handling into one function to get rid of some code duplication.